source/validate_cfg.cpp

```diff
@@ -167,6 +167,153 @@ string ConstructErrorString(const Construct& construct,
+  // uint32_t num_targeting_default = 0u;
+  uint32_t default_case_fall_through = 0u;
+  uint32_t default_target = switch_inst->GetOperandAs<uint32_t>(1u);
+  std::unordered_set<uint32_t> seen;
+  for (uint32_t i = 1; i < switch_inst->operands().size(); i += 2) {
+    uint32_t target = switch_inst->GetOperandAs<uint32_t>(i);
+    if (target == merge->id()) continue;
+
+    if (!seen.insert(target).second) continue;
+
+    // Check each case construct for the following:
+    // 1. It is dominated by the OpSwitch.
+    // 2. It branches to at most one other case construct.
+    // 3. If it branches to another case construct, it must immediately
+    // precede that target (or the fall through of the default case) in the
+    // operand list.
+    bool is_default = i == 1;
+    const auto target_block = function.GetBlock(target).first;
+    if (header->reachable() && target_block->reachable() &&
+        !header->dominates(*target_block)) {
+      return _.diag(SPV_ERROR_INVALID_CFG)
+             << "Selection header " << _.getIdName(header->id())
+             << " does not dominate its case construct " << _.getIdName(target);
+    }
+
+    uint32_t case_fall_through = 0u;
+    if (auto error = FindCaseFallThrough(_, target_block, &case_fall_through,
+                                         merge, case_targets)) {
+      return error;
+    }
+
+    // Track how many time the fall through case has been targeted.
+    if (case_fall_through != 0u) {
+      auto where = num_fall_through_targeted.lower_bound(case_fall_through);
+      if (where == num_fall_through_targeted.end() ||
+          where->first != case_fall_through) {
+        num_fall_through_targeted.insert(where,
+                                         std::make_pair(case_fall_through, 1));
+      } else {
+        where->second++;
+      }
+    }
+
+    if (case_fall_through == default_target) {
+      case_fall_through = default_case_fall_through;
+    }
+    if (case_fall_through != 0u) {
+      if (is_default) {
+        default_case_fall_through = case_fall_through;
+      } else {
+        // Allow code like:
+        // case x:
+        // case y:
+        //   ...
+        // case z:
+        //
+        // Where x and y target the same block and fall through to z.
+        uint32_t j = i;
+        while ((j + 2 < switch_inst->operands().size()) &&
+               target == switch_inst->GetOperandAs<uint32_t>(j + 2)) {
+          j += 2;
+        }
+        if ((switch_inst->operands().size() < j + 2) ||
+            (case_fall_through != switch_inst->GetOperandAs<uint32_t>(j + 2))) {
+          return _.diag(SPV_ERROR_INVALID_CFG)
+                 << "Case construct headed by " << _.getIdName(target)
+                 << " branches to the case construct headed by "
+                 << _.getIdName(case_fall_through)
+                 << ", but does not immediately precede it in the "
+                    "OpSwitch's "
+                    "target list";
```

Comment:
I did this purposefully because the case construct may be many blocks.

Reply:
Oh, so you're trying to say 'one of the blocks in this case construct' instead of just taking the case construct as a whole (regardless of the number of blocks in the case)?

I guess my question would then be, does that matter? Does the word 'construct' when we use it encompass all of the blocks or just something else?

Reply:
I'm talking about the construct by referring to the entry into that construct. It may be pedantic, but many of the issues surrounding structured control flow have been very pedantic so I've gotten into a mode of trying to be more precise. I understand your argument though, not sure which way is best.

Reply:
What about saying something like 'Case construct labeled by'? The term 'headed by' isn't something I'd heard before, whereas a case label, I think, is much more common concept? Is 'headed' what the spec says?